design-proposals/instancetype.kubevirt.io/extend-preferSpread.md

```diff
@@ -0,0 +1,325 @@
+
+This is particularly useful for performance related workloads such as DPDK etc where we would like to make use of SMT on the host in order to map through sibling threads from pCPUs to guest thread siblings. Note that this use case requires `dedicatedCPUPlacement` and `guestMappingPassthrough` to be useful so the CPUManager can allocate complete pCPUs (via the `full-pcpus-only` option) and for these pCPU siblings to be mapped correctly as vCPU thread siblings to the guest.
+
+## Goals
+
+* Allow for vCPUs to be spread across all attributes of the guest visible CPU topology
+
+## User Stories
+
+* As an admin or VM owner I would like to provide preferences that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+* As a VM owner I would like to consume preferences that that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+## Repos
+
+* kubevirt/kubevirt
+
+* kubevirt/common-instancetypes
+
+# Design
+
+### `spec.cpu.spreadOptions`
+
+A new `spreadOptions` struct will be introduced to hold new configurables detailing how vCPUs should be spread over the guest visible CPU topology.
+
+```go
+type spreadAcross string
+
+type SpreadOptions struct {
+	// Across optionally defines how to spread vCPUs across the guest visible topology.
+	// Default: SocketsCores
+	//
+	//+optional
+	Across spreadAcross `json:"across,omitempty"`
+
+	// Ratio optionally defines the ratio to spread vCPUs across the guest visible topology.
+	// Default: 2
+	//
+	//+optional
+	Ratio uint32 `json:"ratio,omitempty"`
+}
+```
+
+The `ratio` option is a replacement for the original `spec.preferSpreadSocketToCoreRatio` configurable. The original being deprecated and discussed later in this proposal.
+
+A new `across` option will be introduced that will map to the following constants and behaviour.
+
+### `spreadAcross` `SocketsCores` (default)
+
+This option will spread vCPUs between sockets and cores. Given the current behaviour of `PreferSpread` this will remain the default.
+
+```go
+const SpreadAcrossSocketsCores SpreadAcross = "SocketsCores"
+```
+
+### `spreadAcross` `CoresThreads`
+
+This option will spread vCPUs between cores and threads with a max of 2 threads per core.
+
+```go
+const SpreadAcrossCoresThreads SpreadAcross = "CoresThreads"
+```
+
+### `spreadAcross` `SocketsCoresThreads`
+
+This option will spread vCPUs between all elements of the CPU topology with a max of 2 threads per core.
+
+```go
+const spreadAcrossSocketsCoresThreads SpreadAcross = "SocketsCoresThreads"
+```
+
+### `spec.preferSpreadSocketToCoreRatio` deprecation
+
+This original configurable for the `preferSpread` `preferredCPUTopology` was added directly to the core preference `spec`.
+
+This was a mistake and given the new nested `spec.cpu.spreadOptions` struct we can now deprecate the original field and move users over to `spec.cpu.spreadOptions.ratio`.
```

Comment:
Deprecation in `v1beta2`, removal in `v1`?

Reply:
I don't think we need a new version to start deprecation of fields?

Reply:
No, also how many steps do you still have in mind before making it v1? :)

Reply:
I'd hope by KubeVirt v1.4 but it all depends on feedback and closing out the remain large gaps like CR upgrades and move to `.Status` etc.

Reply:
ACK